src/libraries/System.Linq/src/System/Linq/Take.cs

```diff
@@ -55,121 +55,133 @@ public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> sourc
             {
                 return startIndex >= endIndex
                     ? Empty<TSource>()
-                    : source.Skip(startIndex).Take(endIndex - startIndex);
+                    : TakeRangeIterator(source, startIndex, endIndex);
             }
 
-            return TakeIterator(source, isStartIndexFromEnd, startIndex, isEndIndexFromEnd, endIndex);
+            return TakeRangeFromEndIterator(source, isStartIndexFromEnd, startIndex, isEndIndexFromEnd, endIndex);
         }
 
-        private static IEnumerable<TSource> TakeIterator<TSource>(
-            IEnumerable<TSource> source, bool isStartIndexFromEnd, int startIndex, bool isEndIndexFromEnd, int endIndex)
+        private static IEnumerable<TSource> TakeRangeFromEndIterator<TSource>(IEnumerable<TSource> source, bool isStartIndexFromEnd, int startIndex, bool isEndIndexFromEnd, int endIndex)
         {
             Debug.Assert(source != null);
+            Debug.Assert(isStartIndexFromEnd || isEndIndexFromEnd);
             Debug.Assert(isStartIndexFromEnd
                 ? startIndex > 0 && (!isEndIndexFromEnd || startIndex > endIndex)
                 : startIndex >= 0 && (isEndIndexFromEnd || startIndex < endIndex));
-            Debug.Assert(endIndex >= 0);
 
-            using IEnumerator<TSource> e = source.GetEnumerator();
-            if (isStartIndexFromEnd)
+            // Attempt to extract the count of the source enumerator,
+            // in order to convert fromEnd indices to regular indices.
+            // Enumerable counts can change over time, so it is very
+            // important that this check happens at enumeration time;
+            // do not move outside of the iterator method.
+            if (source.TryGetNonEnumeratedCount(out int count))
             {
-                if (!e.MoveNext())
+                startIndex = CalculateStartIndex(isStartIndexFromEnd, startIndex, count);
+                endIndex = CalculateEndIndex(isEndIndexFromEnd, endIndex, count);
+
+                if (startIndex >= endIndex)
                 {
                     yield break;
                 }
 
-                int index = 0;
-                Queue<TSource> queue = new();
-                queue.Enqueue(e.Current);
-
-                while (e.MoveNext())
+                foreach (TSource element in TakeRangeIterator(source, startIndex, endIndex))
                 {
-                    checked
-                    {
-                        index++;
-                    }
+                    yield return element;
+                }
```

Comment:
To avoid the extra clutter of more yield breaks, can the `if (startIndex >= endIndex)` condition be inverted?

Reply:
Fixed.